`SideNav` - Documentation (#1249)
hashicorp/design-system: website/docs/components/side-nav/partials/guidelines/guidelines.md

## Usage

### When to use

- To display the primary navigation of the application on a page.

### When not to use

- To move between views within the same context or page, consider [Tabs](#).

## Header

The navigation header displays persistent UI elements that give users quick access to important sections and resources within the application or platform.

### Home link

The home link gives users consistent and quick access to the application's main page or section. Generally, the home or dashboard.

For cloud products, the icon set should always be the HCP service icon. For standalone or open-source products, it should be the product’s service icon (e.g. terraform).

> ![Home link in side-nav's header](/assets/components/side-nav/header-logo.png)

Author: Good suggestion. Added!

### Help dropdown

Use the help dropdown to provide users with access to support and helpful resources that can be easily accessible from anywhere within the application—for example, links to the help center, documentation, or tutorials.

Avoid placing non-help related links or actions, such as user settings or navigation links.

![Help dropdown menu in side-nav's header](/assets/components/side-nav/help-dropdown.png)

### User dropdown

Use the user dropdown to give users quick and easy access to their settings and preferences. The menu should contain links or actions related to the user's profile, settings, and/or preferences. 

Avoid placing links to unrelated pages or actions, such as support or navigation items.

![User dropdown menu in side-nav's header](/assets/components/side-nav/user-dropdown.png)

> 

Reviewer: @didoo curious, will that still be the case once everyone is unified? I feel like our goal has been to represent the most ideal state with _some_ escape hatches when needed, but mostly point to a unified path. WDYT?

> 

Reviewer 2: This is the current sidebar navigation in TFC and the other will be (soon) in Vault. As you can see, it's not always `help` and `user` dropdowns; other products may have other actions. I don't want the documentation to be Cloud-UI-centric.

![TFC](https://user-images.githubusercontent.com/686239/228183090-75af3ed9-c3bd-49dc-9e72-fcae90363089.png)
<img width="286" alt="vault" src="https://user-images.githubusercontent.com/686239/228183097-d06542f2-0139-4773-a033-36fe56a4589a.png">

> 

Author: This paragraph is about the User dropdown specifically. This particular line refers to the content placed inside the dropdown menu. I rephrased the last part to clarify that a little bit more.

## Body

> The body consists of a group of sections with vertical lists of links to the most important parts of the application.

Reviewer: ```suggestion
The body consists of a group of sections with vertical lists of links, typically to the most important parts of the application.
```

This would provide for incorporating @didoo's feedback without too much churn, I think.

Author: Thanks, Melanie. I took your suggestion and rephrased the paragraph a bit, hoping it's clearer now.

> The body consists of a group of sections with vertical lists of links to the most important parts of the application.

Author: Good point. I've added that custom content is supported to that paragraph and an additional section below related to custom content support. Let me know what you think

For cloud products, the icon set should always be the HCP service icon. For standalone or open-source products, it should be the product’s service icon (e.g. terraform).

![Home link in side-nav's header](/assets/components/side-nav/header-logo.png)

### Help dropdown

Use the help dropdown to provide users with access to support and helpful resources that can be easily accessible from anywhere within the application—for example, links to the help center, documentation, or tutorials.

Avoid placing non-help related links or actions, such as user settings or navigation links within the menu.

![Help dropdown menu in side-nav's header](/assets/components/side-nav/help-dropdown.png)

### User dropdown

Use the user dropdown to give users quick and easy access to their settings and preferences. The menu should contain links or actions related to the user's profile, settings, and/or preferences. 

Avoid placing links to unrelated pages or actions, such as support or navigation items within the menu.

![User dropdown menu in side-nav's header](/assets/components/side-nav/user-dropdown.png)

### Custom content

> The `custom` type allows you to support any custom content, local components, or Helios components within the header via an instance swap property (customInstance) in Figma.

Reviewer 3: ```suggestion
The `custom` type supports any custom content, local components, or Helios components within the header via an instance swap property (customInstance) in Figma.
```

![Generic container within the side-nav header](/assets/components/side-nav/custom-header.png)

## Body

The body consists of a group of sections with vertical lists of links, typically to the most important parts of the application. Any custom content or component is also supported by an additional generic container.

### List

#### With title

![Side-nav section with a title](/assets/components/side-nav/section-with-title.png)

#### Without title

![Side-nav section without a title](/assets/components/side-nav/section-without-title.png)

- A title could help users scan the sections and provide context about the links inside each section.

> 

Reviewer 3: ```suggestion
```

- Titles should be meaningful and related to the content within the section.

### List items

#### Without icon

![List items without icons](/assets/components/side-nav/list-item-without-icon.png)

#### With icon

![List items with icons](/assets/components/side-nav/list-item-with-icon.png)

- Use icons to help users recognize and scan the links they are paired with.
- We recommend only using icons in the main or top level navigation.
- Avoid overwriting color styles in icons.

!!! Do

![List items with correct use of icons](/assets/components/side-nav/list-item-with-icon-do.png)
!!!

!!! Dont

![List items with incorrect use of icons](/assets/components/side-nav/list-item-with-icon-dont.png)
!!!

#### With badge

![List item with a badge](/assets/components/side-nav/list-item-with-badge.png)

#### With count

![List item with badge-count](/assets/components/side-nav/list-item-with-count.png)

#### With nested items

> Set `hasSubItems` to show or signify that a link has a nested level of navigation.

Reviewer 3: ```suggestion
Use `hasSubItems` to show or signify that a link has a nested level of navigation.
```
For consistency with the line below

![List item with sub-items](/assets/components/side-nav/list-item-with-nested-items.png)

#### External links

Use `isLinkExternal` to show that the list item is a hyperlink pointing to a page outside the product or platform.

![List item with a external link](/assets/components/side-nav/list-item-with-external-link.png)

!!! Warning

Use external links sparingly. Avoid using this property to link pages that are unrelated to the product's navigation.
!!!

### Custom content

Toggle `hasCustomContent` on to support any additional custom content, local components, or Helios components within the body container via an instance swap property (customInstance) in Figma.

![Generic container within the side-nav body](/assets/components/side-nav/custom-content-body.png)

## Footer

### Context switcher

> The context switcher allows users to switch between different contexts within your product or application. For example, the side navigation can display different links and content based on a particular organization or workspace the user selects.

Reviewer 3: ```suggestion
The context switcher allows users to switch between different contexts within your product or application. For example, the SideNav can display different links and content based on a particular organization or workspace the user selects.
```
Should this specifically reference the component? Open question, not opinionated

Author: @jorytindall I don't think it necessarily should, especially since it's an example to illustrate the purpose and intent of the context switcher. Would something like this sound better?

"The context switcher allows users to switch between different contexts within your product or application. For example, the navigation or application content can change based on a particular organization or workspace the user selects."

Reviewer 3: Yep, I think that reads well, thanks for the clarification!

![Context switcher](/assets/components/side-nav/footer-context-switcher.png)

## Height

> The side navigation should take the full height of the viewport to ensure that the navigation is always visible and accessible to the user.

Author: In this case, yes. It makes sense to reference to the actual component.

![Side-nav height](/assets/components/side-nav/sidenav-height.png)

## Positioning and responsive behaviour

> The SideNav should always be positioned on the left side of the viewport, occupying 100% of the viewport height.

Reviewer 3: Wondering if this and the previous section could be combined, as occupying 100% of the viewport is mentioned twice back-to-back.

Author: Yes. I combined the two sections. Thanks.

> On smaller viewports, the SideNav should collapse to maximize the real estate available on tablet and mobile devices. By tapping the menu icon, users can expand and access the full menu when needed.

Reviewer 3: ```suggestion
On smaller viewports, the SideNav should collapse to maximize the available real estate on tablet and mobile devices. By tapping the menu icon, users can expand and access the full menu when needed.
```
